util/fibers/epoll_socket.cc

```diff
@@ -470,66 +466,72 @@ void EpollSocket::ResetOnRecvHook() {
-        async_read_pending_ = 0;
-      return cb;
-    };
-
-    if (ec) {
-      auto cb = finalize_and_fetch_cb();
-      cb(make_unexpected(ec));
-    } else if (auto res = async_request->Run(native_handle(), is_send); res.first) {
-      auto cb = finalize_and_fetch_cb();
-      cb(res.second);
-    }
-  } else if (recv_hook_registered_ && !is_send && !ec) {
-    on_recv_->cb(RecvNotification{});
-  } else {
-    auto& sync_request = is_send ? write_req_ : read_req_;
-    // It could be that we activated context already, but has not switched to it yet.
-    // Meanwhile a new event has arrived that triggered this callback again.
-    if (sync_request && sync_request->IsSuspended()) {
-      DVSOCK(2) << "Wakey: Schedule read in " << sync_request->name();
-      sync_request->Activate(ec);
+void EpollSocket::HandleAsyncRequest(unsigned epoll_mask) {
+  uint8_t masks[2] = {EpollProactor::EPOLL_IN, EpollProactor::EPOLL_OUT};
+
+  for (unsigned i = 0; i < 2; ++i) {
+    if ((epoll_mask & (masks[i] | kErrMask)) == 0) {
+      continue;
     }
-  }
-}
 
-void EpollSocket::Wakey(uint32_t ev_mask, int error, EpollProactor* cntr) {
-  DVSOCK(2) << "Wakey " << ev_mask;
-#ifdef __linux__
-  constexpr uint32_t kErrMask = EPOLLERR | EPOLLHUP | EPOLLRDHUP;
-#else
-  constexpr uint32_t kErrMask = POLLERR | POLLHUP;
-#endif
+    bool is_send = (i == 1);
+    uint8_t async_pending = is_send ? async_write_pending_ : async_read_pending_;
 
-  error_code ec;
-  if ((ev_mask & POLLERR) && error)
-    ec = error_code{error, system_category()};
-  else if (ev_mask & POLLHUP) {
-    ec = make_error_code(errc::connection_aborted);
-  }
+    error_code ec;
+    bool close_socket = false;
 
-  if (ev_mask & (EpollProactor::EPOLL_IN | kErrMask)) {
-    HandleAsyncRequest(ec, false /*is_send */);
-  }
+    if (epoll_mask & POLLERR) {
+      ec = make_error_code(errc::connection_aborted);
+    } else if (epoll_mask & POLLHUP) {
+      close_socket = true;
+      ec = make_error_code(errc::connection_aborted);
+    }
 
-  if (ev_mask & (EpollProactor::EPOLL_OUT | kErrMask)) {
-    HandleAsyncRequest(ec, true /*is_send */);
+    if (async_pending) {
+      auto& async_request = is_send ? async_write_req_ : async_read_req_;
+      DCHECK(async_request);
+
+      auto finalize_and_fetch_cb = [this, &async_request, is_send]() {
+        auto cb = std::move(async_request->cb);
+        delete async_request;
+        async_request = nullptr;
+        if (is_send)
+          async_write_pending_ = 0;
+        else
+          async_read_pending_ = 0;
+        return cb;
+      };
+
+      // If we got notification, try to run the async request.
+      if (auto res = async_request->Run(native_handle(), is_send); res.first) {
+        auto cb = finalize_and_fetch_cb();
+        cb(res.second);
+      }
+    } else if (recv_hook_registered_ && !is_send) {
+      read_notified_ = 0;
+      on_recv_->cb(close_socket ? RecvNotification{make_error_code(errc::connection_aborted)}
```

Comment:
Does that mean we missed a notification before? 

```
 else if (recv_hook_registered_ && !is_send && !ec) {
    on_recv_->cb(RecvNotification{});
```

Because `!ec` ? So we would miss and not call the on_recv_ callback?